source/common/upstream/outlier_detection_impl.cc

```diff
@@ -229,33 +270,49 @@ void DetectorImpl::onConsecutive5xx(HostSharedPtr host) {
   //          pointer, the detector/cluster must still exist so we can safely fire callbacks.
   //          Otherwise we do nothing since the detector/cluster is already gone.
   std::weak_ptr<DetectorImpl> weak_this = shared_from_this();
-  dispatcher_.post([weak_this, host]() -> void {
+  dispatcher_.post([weak_this, host, type]() -> void {
     std::shared_ptr<DetectorImpl> shared_this = weak_this.lock();
     if (shared_this) {
-      shared_this->onConsecutive5xxWorker(host);
+      shared_this->onConsecutiveErrorWorker(host, type);
     }
   });
 }
 
-void DetectorImpl::onConsecutive5xxWorker(HostSharedPtr host) {
-  // This comes in cross thread. There is a chance that the host has already been removed from
+void DetectorImpl::onConsecutive5xx(HostSharedPtr host) {
+  notifyMainThreadConsecutiveError(host, EjectionType::Consecutive5xx);
+}
+
+void DetectorImpl::onConsecutiveGatewayFailure(HostSharedPtr host) {
+  notifyMainThreadConsecutiveError(host, EjectionType::ConsecutiveGatewayFailure);
+}
+
+void DetectorImpl::onConsecutiveErrorWorker(HostSharedPtr host, EjectionType type) {
+  // Ejections come in cross thread. There is a chance that the host has already been removed from
   // the set. If so, just ignore it.
   if (host_monitors_.count(host) == 0) {
     return;
   }
-
   if (host->healthFlagGet(Host::HealthFlag::FAILED_OUTLIER_CHECK)) {
     return;
   }
 
-  stats_.ejections_consecutive_5xx_.inc();
-  ejectHost(host, EjectionType::Consecutive5xx);
-
-  // Reset the DetectorHostMonitor's consecutive_5xx_ counter. The reset is performed here
-  // on the onConsecutive5xxWorker call to prevent thread thrashing. The consecutive_5xx_
-  // counter needs to be reset in order to allow the monitor to detect a bout of consecutive
-  // 5xx responses even if the monitor is not charged with an interleaved non-5xx code.
-  host_monitors_[host]->resetConsecutive5xx();
+  // We also reset the appropriate counter here to allow the monitor to detect a bout of consecutive
+  // error responses even if the monitor is not charged with an interleaved non-error code.
+  switch (type) {
+  case EjectionType::Consecutive5xx:
+    stats_.ejections_consecutive_5xx_.inc(); // Deprecated
+    stats_.ejections_detected_consecutive_5xx_.inc();
+    ejectHost(host, EjectionType::Consecutive5xx);
+    host_monitors_[host]->resetConsecutive5xx();
+    break;
+  case EjectionType::ConsecutiveGatewayFailure:
+    stats_.ejections_detected_consecutive_gateway_failure_.inc();
+    ejectHost(host, EjectionType::ConsecutiveGatewayFailure);
+    host_monitors_[host]->resetConsecutiveGatewayFailure();
+    break;
+  default:
```

Comment:
nit: is this default case needed? Compiler should be OK with all cases implemented.